src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1407,32 +1407,53 @@ abstract class RefChecks extends Transform {
         false
     }
 
-    private def checkTypeRef(tp: Type, tree: Tree, skipBounds: Boolean): Unit = tp match {
-      case TypeRef(pre, sym, args) =>
-        tree match {
-          case tt: TypeTree if tt.original == null => // scala/bug#7783 don't warn about inferred types
-                                                      // FIXME: reconcile this check with one in resetAttrs
-          case _ => checkUndesiredProperties(sym, tree.pos)
+    private object RefCheckTypeMap extends TypeMap {
+      object ExistentialToWildcard extends TypeMap {
+        override def apply(tpe: Type): Type =
+          if (tpe.typeSymbol.isExistential) WildcardType else tpe.mapOver(this)
+      }
+
+      private[this] var skipBounds = false
+      private[this] var tree: Tree = EmptyTree
+
+      def check(tpe: Type, tree: Tree): Type = {
+        this.tree = tree
+        try apply(tpe) finally {
+          skipBounds = false
+          this.tree = EmptyTree
         }
-        if (sym.isJavaDefined)
-          sym.typeParams foreach (_.cookJavaRawInfo())
-        if (!tp.isHigherKinded && !skipBounds)
-          checkBounds(tree, pre, sym.owner, sym.typeParams, args)
-      case _ =>
-    }
+      }
 
-    private def checkTypeRefBounds(tp: Type, tree: Tree) = {
-      var skipBounds = false
-      tp match {
-        case AnnotatedType(ann :: Nil, underlying) if ann.symbol == UncheckedBoundsClass =>
+      // check all bounds, except those that are existential type parameters
+      // or those within typed annotated with @uncheckedBounds
+      override def apply(tpe: Type): Type = tpe match {
+        case tpe: AnnotatedType if tpe.hasAnnotation(UncheckedBoundsClass) =>
+          // scala/bug#7694 Allow code synthesizers to disable checking of bounds for TypeTrees based on inferred LUBs
+          // which might not conform to the constraints.
+          val savedSkipBounds = skipBounds
           skipBounds = true
-          underlying
+          try tpe.mapOver(this).filterAnnotations(_.symbol != UncheckedBoundsClass)
+          finally skipBounds = savedSkipBounds
+        case tpe: TypeRef =>
+          checkTypeRef(ExistentialToWildcard(tpe))
```

Comment:
This used to be discarded:

```scala
scala> trait T[X <: String]
trait T

scala> class C { def f: T[X forSome { type X <: Int }] = ??? }
                            ^
       warning: the existential type X forSome { type X <: Int }, which cannot be expressed by wildcards, should be enabled
       by making the implicit value scala.language.existentials visible.
                        ^
       error: type arguments [X forSome { type X <: Int }] do not conform to trait T's type parameter bounds [X <: String]
```

now it passes. OTOH, these two passed before (2.13.6/7):

```scala
scala> class C { def f: T[_ <: Int] = ??? }
class C

scala> class C { def f: T[X] forSome { type X <: Int } = ??? }
class C
```

WDYT?

Reply:
I see now, this is basically the oversight in the old impl, which only collected existentials at the top level to replace them with wildcards.

I guess to be more precise, we should only replace existentials with wildcards that were originally actual wildcards in source? But that's not possible to tell?

Reply:
Ahh I see what you mean - let me think about it some more. We want to convert existential quantifiers that come from the outside to wildcards, but not those that are nested within.

> I guess to be more precise, we should only replace existentials with wildcards that were originally actual wildcards in source? But that's not possible to tell?

It's possible to tell, but that actually makes some fail. As you noticed existential quantifiers already get a pass when it comes to bounds regardless of whether they come from wildcards or not. Do you think that was intentional? If not, we can make that case fail and it would solve our other problem as well.

Reply:
I really don't *know*, I'm just trying to infer stuff from intuition / experience.

To start with the basics, I guess `T[_]` should always pass for practical reasons? I can't find anything about it in the spec.

> wildcard type is of the form _>:𝐿<:𝑈. If >:𝐿 is missing, >:scala.Nothing is assumed, if <:𝑈 is missing, <:scala.Any is assumed. A wildcard type is a shorthand for an existentially quantified type variable

> Say the type parameters have lower bounds 𝐿1,…,𝐿𝑛 and upper bounds 𝑈1,…,𝑈𝑛. The parameterized type is well-formed if each actual type parameter conforms to its bounds, i.e. 𝐿𝑖<:𝑇𝑖<:𝑈𝑖

Is checking the bounds required for type safety? Or will mismatching bounds be always flagged in subtype tests?

```
scala> trait T[X <: String] { def x: X }
trait T

scala> def f(t: T[_ <: Int]) = t.x
def f(t: T[_ <: Int]): Int

scala> def t: T[String] = ???
def t: T[String]

scala> f(t)
         ^
       error: type mismatch;
        found   : T[String]
        required: T[_ <: Int]
```

Reply:
Scala 3 also seems to accept wildcards with non-matching bounds

```
scala> trait T[X <: String]
// defined trait T

scala> class C { def f(x: T[_ <: Int]): T[_ <: Int] = x }
// defined class C
```

So let's go with that.

Reply:
I will see if we need a followup for `class C { def f: T[X forSome { type X <: Int }] = ??? }` because it still looks wrong to me. `X forSome { type X <: Int }` is not a subtype of `String` although in `T[X] forSome { type X <: Int }` we accept `X` as a subtype of `String`. Good idea to check Scala 3 - but of course it doesn't have existential types.